### PR TITLE
feat(Stream): zipLatestAll

### PR DESCRIPTION
**Pull request by dilame**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR is based on my another PR https://github.com/Effect-TS/effect/pull/2830 so it is better to resolve it first

## Related

- Related Issue #

- Closes https://github.com/Effect-TS/effect/issues/2718

### CODE REVIEW

File: packages/effect/src/Stream.ts [hunk 1]
```diff
@@ -4329,6 +4329,44 @@ export const zipLatest: {
...
+ * // Output:
+ * [ 0, 0, 0 ]
+ * [ 1, 0, 0 ]
+ * [ 1, 1, 0 ]
+ * [ 2, 1, 0 ]
+ * [ 3, 1, 0 ]
+ * [ 3, 1, 1 ]
+ * // .....
+ *
+ * @since 2.0.0
```
Comment (tim-smart) [inline packages/effect/src/Stream.ts:4359]:
```suggestion
 * @since 3.3.0
```

File: packages/effect/src/internal/stream.ts [hunk 2]
```diff
@@ -7565,6 +7566,34 @@ export const zipLatest = dual<
...
+  if (Arr.isTupleOfAtLeast(streams, 2)) {
+    const [head, ...tail] = streams
+    return zipLatestWith(
+      head,
+      zipLatestAll(...tail),
+      (first, second) => [first, ...second] as ZipLatestAllValues<T>
+    )
+  }
+  return empty
+}
```
Comment (tim-smart) [inline packages/effect/src/internal/stream.ts:7595]: I think just using .length is easier to read and maintain.

```suggestion
): Stream.Stream<ZipLatestAllValues<T>, ZipLatestAllErrors<T>, ZipLatestAllContext<T>> => {
  if (streams.length === 0) {
    return empty
  } else if (streams.length === 1) {
    return map(streams[0], (x) => [x] as ZipLatestAllValues<T>))
  }
  const [head, ...tail] = streams
  return zipLatestWith(
    head,
    zipLatestAll(...tail),
    (first, second) => [first, ...second] as ZipLatestAllValues<T>
  )
}
```

File: packages/effect/src/Stream.ts [hunk 3]
```diff
@@ -4329,6 +4329,44 @@ export const zipLatest: {
...
+ * @category zipping
+ */
+export const zipLatestAll: <T extends Array<Stream<unknown>>>(
```
Comment (tim-smart) [inline packages/effect/src/Stream.ts:4362]:
```suggestion
export const zipLatestAll: <T extends ReadonlyArray<Stream<any, any, any>>>(
```

Otherwise you will run into assignability issues.

File: packages/effect/src/Stream.ts [hunk 4]
```diff
@@ -4329,6 +4329,44 @@ export const zipLatest: {
...
+  ...streams: T
+) => Stream<
+  internal.ZipLatestAllValues<T>,
```
Comment (tim-smart) [inline packages/effect/src/Stream.ts:4365]: Type helpers should not be internal, as they are stripped from the final build.

Also, I think these types should just be inlined :)

```suggestion
  T extends ReadonlyArray<Stream<infer _R, infer _E, infer _A>> ? _R : never,
```

Will need to test if this works with empty arguments too.

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: packages/effect/src/internal/stream.ts (hunk 2 above)
Comment (dilame) [reply packages/effect/src/internal/stream.ts:7595]: If you are checking only length of the array you will not get type narrowing, so 

```ts
if(streams.length === 1) {
  streams[0]
  // ^? number | undefined
}
```

My suggestion is 

```ts
if(Preficate.isTupleOf(streams, 1)) {
  streams[0]
  // ^? number
}
```

My PR with `isTupleOf` has already been merged into `next-minor` (https://github.com/Effect-TS/effect/commit/3182c89e57c87c18dbe66271149ca6a23a5b3b1b), so i think it's a perfect solution here:)

File: packages/effect/src/Stream.ts (hunk 4 above)
Comment (dilame) [reply packages/effect/src/Stream.ts:4365]: This is the suggestion that i don't understand, because AFAIU
`T extends ReadonlyArray<Stream<infer _R, infer _E, infer _A>> ? _A : never,` is not the same as `ZipLatestAllValues`.

I am also not sure what to do with the type helpers, because if we choose to inline it, we would need to do it in 2 both -`Stream.ts` and `internal/stream.ts`. These types will look pretty heavy inlined. Is it ok? What do you think about it?

File: packages/effect/src/internal/stream.ts [hunk 5]
```diff
@@ -7565,6 +7566,34 @@ export const zipLatest = dual<
...
+  Array<Stream.Stream<unknown, infer E, unknown>> ? E : never
+/** @internal */
+export type ZipLatestAllContext<T extends Array<Stream.Stream<unknown, unknown, unknown>>> = T extends
+  Array<Stream.Stream<unknown, unknown, infer R>> ? R : never
+
+export const zipLatestAll = <T extends Array<Stream.Stream<unknown>>>(
+  ...streams: T
+): Stream.Stream<ZipLatestAllValues<T>, ZipLatestAllErrors<T>, ZipLatestAllContext<T>> => {
+  if (Arr.isTupleOf(streams, 1)) {
+    return pipe(streams[0], map((x) => [x] as ZipLatestAllValues<T>))
```
Comment (dilame) [inline packages/effect/src/internal/stream.ts:7584]: BTW, @tim-smart , i also use `ZipLatestAllValues` type helper in 2 places inside the function (_here and 7 lines below from here_) for manual type-casting, because i can not think of a way to get rid of it. 
In total, `ZipLatestAllValues` is used in 4 places. Should i  still inline it?

**Comment by dilame** [comment]
Seems like i did something wrong with merging:) Should i open a new PR?

File: packages/effect/src/internal/stream.ts (hunk 5 above)
Comment (tim-smart) [reply packages/effect/src/internal/stream.ts]: Yes still inline it :)

**Comment by tim-smart** [comment]
> Seems like i did something wrong with merging:) Should i open a new PR?

Rebased and added some tests. There is one type level test that needs fixing.

**Comment by dilame** [comment]
Seems like i fixed the problem with extra commits, and also fixed build errors, should be ok now

**Comment by dilame** [comment]
I can see that pages build is failed, but the reason is
`Error: The operation was canceled.`
I'm not sure what to do in this case

File: packages/effect/src/Stream.ts [hunk 6]
```diff
@@ -4350,6 +4350,45 @@ export const zipLatest: {
...
+ *
+ * Effect.runPromise(Stream.runDrain(stream))
+ * // Output:
+ * // [ 0, 0, 0 ]
+ * // [ 1, 0, 0 ]
+ * // [ 1, 1, 0 ]
+ * // [ 2, 1, 0 ]
+ * // [ 3, 1, 0 ]
+ * // [ 3, 1, 1 ]
+ * // .....
```
Comment (IMax153) [inline packages/effect/src/Stream.ts:4378]: @dilame - I believe the reason docs generation is failing is because this example (which is run by our documentation generator to ensure validity` creates a stream that never terminates. Can you please update the example to the following:

```suggestion
 * @example
 * import { Stream, Schedule, Console, Effect } from "effect"
 *
 * const stream = Stream.zipLatestAll(
 *     Stream.fromSchedule(Schedule.spaced('5 millis')),
 *     Stream.fromSchedule(Schedule.spaced('10 millis')),
 *     Stream.fromSchedule(Schedule.spaced('20 millis')),
 * ).pipe(Stream.take(6), Stream.tap(Console.log))
 *
 * Effect.runPromise(Stream.runDrain(stream))
 * // Output:
 * // [ 0, 0, 0 ]
 * // [ 1, 0, 0 ]
 * // [ 1, 1, 0 ]
 * // [ 2, 1, 0 ]
 * // [ 3, 1, 0 ]
 * // [ 3, 1, 1 ]
 * // .....
```

**Review by tim-smart — APPROVED** [review]

**Review by IMax153 — APPROVED** [review]
